### PR TITLE
Issue #366 - Literal suffixes should be upper case

### LINKED ISSUE #366: Literal suffixes should be upper case
```
var x = 10l; //Noncompliant, is it 101 or 10? It's hard to read.
var y = 10L;
```

R# ID: LongLiteralEndingLowerL

https://jira.sonarsource.com/browse/RSPEC-818

### PR DESCRIPTION
**Pull request by tvsonar**
Fixes #366

### CODE REVIEW

File: src/SonarLint/Rules.Description/S818.html [hunk 1]
```diff
@@ -0,0 +1,33 @@
+<p>
+    Using upper case literal suffixes removes the potential ambiguity between "1" (digit 1) and "l" (letter el) for declaring literals.
```
Comment (duncanpMS) [inline src/SonarLint/Rules.Description/S818.html]: This doesn't explain why "u", "d", and "f" should be upper-case
e.g. "For consistency, upper-case all literal suffixes."

File: src/SonarLint/Rules/LiteralSuffixUpperCase.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    [Rule(DiagnosticId, RuleSeverity, Title, IsActivatedByDefault)]
+    [Tags(Tag.Cert, Tag.Convention, Tag.Misra, Tag.Pitfall)]
+    public class LiteralSuffixUpperCase : DiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S818";
+        internal const string Title = "Literal suffixes should be upper case";
+        internal const string Description =
+            "Using upper case literal suffixes removes the potential ambiguity between \"1\" (digit 1) and \"l\" " +
+            "(letter el) for declaring literals.";
+        internal const string MessageFormat = "Upper-case this literal suffix.";
```
Comment (duncanpMS) [inline src/SonarLint/Rules/LiteralSuffixUpperCase.cs]: Consistency: this "upper-case" is hyphenated, but other occurrences are not. I'd hyphenate them all.

File: src/SonarLint/Rules/LiteralSuffixUpperCase.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+                    var text = literal.Token.Text;
+
+                    if (text.StartsWith("0x", System.StringComparison.InvariantCultureIgnoreCase))
+                    {
+                        return;
+                    }
+
+                    var reversedTextEnding = new string(text.Reverse().TakeWhile(char.IsLetter).ToArray());
+
+                    if (reversedTextEnding != reversedTextEnding.ToUpper())
```
Comment (duncanpMS) [inline src/SonarLint/Rules/LiteralSuffixUpperCase.cs]: ToUpperInvariant()?

File: src/SonarLint/Rules/LiteralSuffixUpperCaseCodeFixProvider.cs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+
+namespace SonarLint.Rules
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp)]
+    public class LiteralSuffixUpperCaseCodeFixProvider : CodeFixProvider
+    {
+        public const string Title = "Make literal uppercase";
```
Comment (duncanpMS) [inline src/SonarLint/Rules/LiteralSuffixUpperCaseCodeFixProvider.cs]: It probably needs the word "suffix" in there somewhere e.g.
"Make suffix upper-case" / "Make literal suffix upper-case"

File: src/Tests/SonarLint.UnitTest/TestCases/LiteralSuffixUpperCase.Fixed.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+            const ulong c = 0UL;
+            const ulong d = 0UL;
+            const decimal e = 1.2M;
+            const float f = 1.2F;
+            const double g = 1.2D;
+            const double h = 1.2E-10;
+            const int i = 0xF;
+        }
+    }
+}
```
Comment (duncanpMS) [inline src/Tests/SonarLint.UnitTest/TestCases/LiteralSuffixUpperCase.Fixed.cs]: Additional test cases:
* uppercase hex: i.e. 0Xa
* signed constants

**Comment by duncanpMS** [comment]
LGTM

File: src/SonarLint/Rules/LiteralSuffixUpperCase.cs (hunk 2 above)
Comment (tvsonar) [inline src/SonarLint/Rules/LiteralSuffixUpperCase.cs]: I raised these points to @ganncamp.

Comment (ganncamp) [inline src/SonarLint/Rules/LiteralSuffixUpperCase.cs]: I've removed the hyphen in the rspec. IMO, should probably be "uppercase" in all cases except the message ("upper case" there) for the same reason that you "Log in with your login". But both forms seem to be acceptable, so I'll leave it as is.

**Pull request merged**